### PR TITLE
`simple_query_string` implementation.

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
Implementation for `simple_query_string` copied from `dev-simple_query_string-#192-impl2`, since that branch was corrupted by incorrect merge.
See also #61.
All PR review notes have been addressed.

Signed-off-by: Yury Fridlyand <yuryf@bitquilltech.com>

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by forestmvey — APPROVED** [review]

File: integ-test/src/test/java/org/opensearch/sql/sql/SimpleQueryStringIT.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    assertNotEquals(0, result.getInt("total"));
+  }
+
+  @Test
+  public void verify_wildcard_test() throws IOException {
+    String query1 = "SELECT count(*) FROM "
+        + Index.BEER.getName() + " WHERE simple_query_string(['Tags'], 'taste')";
+    var result1 = new JSONObject(executeQuery(query1, "jdbc"));
+    String query2 = "SELECT count(*) FROM "
+        + Index.BEER.getName() + " WHERE simple_query_string(['T*'], 'taste')";
```
Comment (MaxKsyunz) [inline integ-test/src/test/java/org/opensearch/sql/sql/SimpleQueryStringIT.java:41]: Couldn't `TestConstants.TEST_INDEX_BEER` be used here?

File: sql/src/main/antlr/OpenSearchSQLParser.g4 [hunk 2]
```diff
@@ -318,9 +318,20 @@ specificFunction
...
+// Field is a single column
+singleFieldRelevanceFunction
+    : singleFieldRelevanceFunctionName LR_BRACKET
+        field=relevanceField COMMA query=relevanceQuery
+        (COMMA relevanceArg)* RR_BRACKET;
+
+// Field is a list of columns
+multiFieldRelevanceFunction
+    : multiFieldRelevanceFunctionName LR_BRACKET
+        LT_SQR_PRTHS field=relevanceFieldAndWeight (COMMA field=relevanceFieldAndWeight)* RT_SQR_PRTHS
```
Comment (MaxKsyunz) [inline sql/src/main/antlr/OpenSearchSQLParser.g4:333]: I believe if we add a separate rule for field list (everything between square brackets), we can simplify AstExpressionBuilder.

File: sql/src/main/java/org/opensearch/sql/sql/parser/AstExpressionBuilder.java [hunk 3]
```diff
@@ -362,9 +363,17 @@ public UnresolvedExpression visitConvertedDataType(
 
   @Override
   public UnresolvedExpression visitRelevanceFunction(RelevanceFunctionContext ctx) {
-    return new Function(
-        ctx.relevanceFunctionName().getText().toLowerCase(),
-        relevanceArguments(ctx));
+    if (ctx.singleFieldRelevanceFunction() != null) {
```
Comment (MaxKsyunz) [inline sql/src/main/java/org/opensearch/sql/sql/parser/AstExpressionBuilder.java:366]: Similar comments I made for PPL AstExpressionBuilder apply here.

File: core/src/main/java/org/opensearch/sql/ast/expression/RelevanceFieldList.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import org.opensearch.sql.ast.AbstractNodeVisitor;
+
+/**
+ * Expression node that includes a list of RelevanceField nodes.
+ */
+@EqualsAndHashCode(callSuper = false)
+@AllArgsConstructor
+public class RelevanceFieldList extends UnresolvedExpression {
+  @Getter
+  private java.util.Map<UnresolvedExpression, UnresolvedExpression> fieldList;
```
[the remark this answers is not in the document]
Comment (acarbonetto) [reply core/src/main/java/org/opensearch/sql/ast/expression/RelevanceFieldList.java:23]: We talked about using the Field class to store the string/float parameters.  If that doesn't work, we should consider creating an Object to store this information.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/SimpleQueryStringQuery.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+import org.opensearch.index.query.SimpleQueryStringBuilder;
+import org.opensearch.index.query.SimpleQueryStringFlag;
+import org.opensearch.sql.data.model.ExprValue;
+import org.opensearch.sql.exception.SemanticCheckException;
+import org.opensearch.sql.expression.Expression;
+import org.opensearch.sql.expression.FunctionExpression;
+import org.opensearch.sql.expression.NamedArgumentExpression;
+import org.opensearch.sql.opensearch.storage.script.filter.lucene.LuceneQuery;
+
+public class SimpleQueryStringQuery extends LuceneQuery {
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/lucene/relevance/SimpleQueryStringQuery.java]: Fixed in 521beca.

File: opensearch/src/test/java/org/opensearch/sql/opensearch/storage/script/filter/FilterQueryBuilderTest.java [hunk 6]
```diff
@@ -386,6 +424,65 @@ void should_build_match_phrase_query_with_default_parameters() {
                 dsl.namedArgument("query", literal("search query")))));
   }
 
+  @Test
+  // Test is disabled because DSL because of:
+  // 1) DSL reverses order of the `fields`
+  // 2) `flags` are printed by OpenSearch (not by the plugin) as an integer
+  // 3) parameters are printed -//- not in lexicographical order
+  @Disabled
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply opensearch/src/test/java/org/opensearch/sql/opensearch/storage/script/filter/FilterQueryBuilderTest.java]: Fixed in 2ac159f.

File: ppl/src/main/java/org/opensearch/sql/ppl/parser/AstExpressionBuilder.java [hunk 7]
```diff
@@ -252,9 +254,17 @@ public UnresolvedExpression visitConvertedDataType(ConvertedDataTypeContext ctx)
 
   @Override
   public UnresolvedExpression visitRelevanceExpression(RelevanceExpressionContext ctx) {
-    return new Function(
-        ctx.relevanceFunctionName().getText().toLowerCase(),
-        relevanceArguments(ctx));
+    if (ctx.singleFieldRelevanceFunction() != null) {
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply ppl/src/main/java/org/opensearch/sql/ppl/parser/AstExpressionBuilder.java]: Fixed in 0817d9e.

File: ppl/src/main/java/org/opensearch/sql/ppl/parser/AstExpressionBuilder.java [hunk 8]
```diff
@@ -252,9 +254,17 @@ public UnresolvedExpression visitConvertedDataType(ConvertedDataTypeContext ctx)
...
+      return new Function(
+          ctx.singleFieldRelevanceFunction()
+              .singleFieldRelevanceFunctionName().getText().toLowerCase(),
+          singleFieldRelevanceArguments(ctx.singleFieldRelevanceFunction()));
+    } else {
+      return new Function(
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply ppl/src/main/java/org/opensearch/sql/ppl/parser/AstExpressionBuilder.java]: Fixed in 0817d9e.

File: opensearch/src/test/java/org/opensearch/sql/opensearch/storage/script/filter/FilterQueryBuilderTest.java [hunk 9]
```diff
@@ -367,6 +370,41 @@ void match_invalid_parameter() {
     assertEquals("Parameter invalid_parameter is invalid for match function.", msg);
   }
 
+  @Test
+  // `buildQuery` calls OpenSearch's (not plugin's) `toString()` function which prints
+  // - fields' weights inside quotes
+  // - `flags` as an integer, but not as an enum
+  // - parameters not in alphabetical order
+  @Disabled
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply opensearch/src/test/java/org/opensearch/sql/opensearch/storage/script/filter/FilterQueryBuilderTest.java]: Fixed in 2ac159f.

File: core/src/main/java/org/opensearch/sql/analysis/ExpressionAnalyzer.java [hunk 10]
```diff
@@ -158,6 +163,18 @@ public Expression visitAggregateFunction(AggregateFunction node, AnalysisContext
...
 
+  @Override
+  public Expression visitRelevanceFieldList(RelevanceFieldList node, AnalysisContext context) {
+    return new LiteralExpression(new ExprTupleValue(new LinkedHashMap<>(node
+        .getFieldList()
+        .entrySet()
+        .stream()
+        .collect(ImmutableMap.toImmutableMap(
+            n -> n.getKey().toString(),
+            n -> ExprValueUtils.floatValue((Float) ((Literal) n.getValue()).getValue())
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/analysis/ExpressionAnalyzer.java]: Fixed in 015aa06.

File: integ-test/src/test/java/org/opensearch/sql/sql/SimpleQueryStringIT.java (hunk 1 above)
Comment (Yury-Fridlyand) [reply integ-test/src/test/java/org/opensearch/sql/sql/SimpleQueryStringIT.java]: Fixed in c1a91fc.

**Review by MaxKsyunz — APPROVED** [review]

File: opensearch/src/test/java/org/opensearch/sql/opensearch/storage/script/filter/FilterQueryBuilderTest.java [hunk 11]
```diff
@@ -476,6 +592,52 @@ void match_phrase_too_many_args() {
...
+  @Test
+  void simple_query_string_missing_query() {
+    var msg = assertThrows(ExpressionEvaluationException.class, () ->
+        dsl.simple_query_string(
+            dsl.namedArgument("fields", DSL.literal(
+                new ExprTupleValue(new LinkedHashMap<>(ImmutableMap.of(
+                    "field1", ExprValueUtils.floatValue(1.F),
+                    "field2", ExprValueUtils.floatValue(.3F)))))))).getMessage();
+    assertEquals("simple_query_string function expected {[STRUCT,STRING],[STRUCT,STRING,STRING],"
+          + "[STRUCT,STRING,STRING,STRING],[STRUCT,STRING,STRING,STRING,STRING],[STRUCT,STRING,"
```
Comment (acarbonetto) [inline opensearch/src/test/java/org/opensearch/sql/opensearch/storage/script/filter/FilterQueryBuilderTest.java:623]: 😮

File: ppl/src/main/antlr/OpenSearchPPLLexer.g4 [hunk 12]
```diff
@@ -141,6 +141,13 @@ SINGLE_QUOTE:                       '\'';
 DOUBLE_QUOTE:                       '"';
 BACKTICK:                           '`';
 
+// Operators. Bit
+
+BIT_NOT_OP:                         '~';
+//BIT_OR_OP:                          '|';
```
Comment (acarbonetto) [inline ppl/src/main/antlr/OpenSearchPPLLexer.g4:147]: commented code can be removed

File: ppl/src/main/antlr/OpenSearchPPLLexer.g4 [hunk 13]
```diff
@@ -265,19 +272,40 @@ IF:                                 'IF';
...
 MINIMUM_SHOULD_MATCH:               'MINIMUM_SHOULD_MATCH';
+OPERATOR:                           'OPERATOR';
+PHRASE_SLOP:                        'PHRASE_SLOP';
+PREFIX_LENGTH:                      'PREFIX_LENGTH';
+QUOTE_ANALYZER:                     'QUOTE_ANALYZER';
+QUOTE_FIELD_SUFFIX:                 'QUOTE_FIELD_SUFFIX';
+REWRITE:                            'REWRITE';
+SLOP:                               'SLOP';
+TIE_BREAKER:                        'TIE_BREAKER';
+//TIME_ZONE:                          'TIME_ZONE'; // already defined on line 63
```
Comment (acarbonetto) [inline ppl/src/main/antlr/OpenSearchPPLLexer.g4:305]: commented code can be removed

**Review by acarbonetto — APPROVED** [review]

File: ppl/src/main/antlr/OpenSearchPPLLexer.g4 (hunk 12 above)
Comment (Yury-Fridlyand) [reply ppl/src/main/antlr/OpenSearchPPLLexer.g4]: Fixed in b3bca7c4.

File: ppl/src/main/antlr/OpenSearchPPLLexer.g4 (hunk 13 above)
Comment (Yury-Fridlyand) [reply ppl/src/main/antlr/OpenSearchPPLLexer.g4]: Fixed in b3bca7c4.

**Pull request merged**